resources/Microsoft.DotNet.Dsc/Microsoft.DotNet.Dsc.psm1

```diff
@@ -0,0 +1,547 @@
+    )
+
+    try
+    {
+        Invoke-Expression "& `"$DotNetCliPath`" $Command"
+    }
+    catch
+    {
+        throw "Executing dotnet.exe with {$Command} failed."
+    }
+}
+
+# Keeps the path of the code.exe CLI path.
+$DotNetCliPath = Get-DotNetPath
+
+#endregion Functions
+
+#region Classes
+<#
+.SYNOPSIS
+    This class is used to install and uninstall .NET SDK tools globally or use the tool path directory.
+#>
+[DSCResource()]
+class DotNetToolPackage
+{
+    [DscProperty(Key)]
+    [string] $PackageId
+
+    [DscProperty()]
+    [string] $Version
+
+    [DscProperty()]
+    [string[]] $Commands
+
+    [DscProperty()]
+    [bool] $PreRelease = $false
+
+    [DscProperty()]
+    [string] $ToolPath
+
+    [DscProperty()]
+    [bool] $Exist = $true
+
+    static [hashtable] $InstalledPackages
+
+    DotNetToolPackage()
+    {
+        [DotNetToolPackage]::GetInstalledPackages()
+    }
+
+    DotNetToolPackage([string] $PackageId, [string] $Version, [string[]] $Commands, [bool] $PreRelease, [string] $ToolPath, [bool] $Exist)
+    {
+        $this.PackageId = $PackageId
+        $this.Version = $Version
+        $this.Commands = $Commands
+        $this.PreRelease = $PreRelease
+        $this.ToolPath = $ToolPath
+        $this.Exist = $Exist
+    }
+
+    [DotNetToolPackage] Get()
+    {
+        # get the properties of the object currently set
+        $properties = $this.ToHashTable()
+
+        # refresh installed packages
+        [DotNetToolPackage]::GetInstalledPackages($properties)
+
+        # get the current state
+        $currentState = [DotNetToolPackage]::InstalledPackages[$this.PackageId]
+
+        # get the properties of the object currently set
+        $properties = $this.ToHashTable()
+
+        # refresh installed packages
+        [DotNetToolPackage]::GetInstalledPackages($properties)
+
+        # current state
+        $currentState = [DotNetToolPackage]::InstalledPackages[$this.PackageId]
+
+        if ($null -ne $currentState)
+        {
+            if ($this.Version -and ($this.Version -ne $currentState.Version))
+            {
+                # See treatment: https://learn.microsoft.com/en-us/nuget/concepts/package-versioning?tabs=semver20sort#normalized-version-numbers
+                # in this case, we misuse revision if beta,alpha, rc are present and grab the highest revision
+                $installedVersion = Get-Semver -Version $currentState.Version
+                $currentVersion = Get-Semver -Version $this.Version
+                if ($currentVersion -gt $installedVersion)
+                {
+                    $currentState.Exist = $false
+                    $currentState.Version = $this.Version
+                }
+                else 
+                {
+                    # Don't want to downgrade scenarios
+                    Throw "The package version cannot be less than the installed version. Please provide a version greater than or equal to the installed version."
+                }
```

Comment:
We shouldn't be throwing on a Get() call. I also believe that we should still support downgrade scenarios even if this is not a common scenario. DSC should be able to set any desired state even if that involves downgrading.

Reply:
Please check once again. The logic is a bit different. Check out the GitHub issue linked at the function. Not all SDKs have the downgrade option ;)